File: examples/kotlin/src/main/kotlin/adr/spine/boundary/Action.kt
// ── spine/boundary/action — the name→ToolResult map (F1) ───────────────────
// The human path into the fold used to be unspecified: fold consumed [ToolResult],
// the surface emitted Action, and NOTHING named the conversion. This file is that
// conversion, and it is the symmetric twin of 6.8's name→Command map — both fed by
// ONE registration.
//
//     ActionResolution(registry).resolve(action, ctx) -> ToolResult   name → ToolResult
//     ActionResolution(registry).sign(result, sig, id) -> Command      name → Command
//
// Both are members of a CONSTRUCTED type, never top-level functions. A top-level
// function has no instance: nothing builds it, so nothing can stand in for it, so it
// can only be reached through whatever calls it. The registry is the state the pair
// shares, so it is constructor-held and drops out of both signatures.
//
// The Boundary builds this itself rather than receiving it injected. That is deliberate:
// an injected resolver would let a test mint ToolResults from a second site, which is
// exactly what C7 below forbids. Instantiated, therefore testable; not bound, therefore
// still the single production site.
//
// Because the open-name guard lives HERE — at the boundary, where the open name
// actually arrives — the fold has no `else` arm at all. It is exhaustive over a
// fully sealed ToolResult including Unhandled and Refused. That is 6.10's "close
// what you own; guard what you do not" put in the right place, and it is what makes
// F10's compile-time edit list total.
//
// This is also the SINGLE PRODUCTION SITE of every ToolResult in the system
// (gate check C7), so a recorded result can never disagree with what was folded.

package adr.spine.boundary

import adr.contract.Command
import adr.contract.ToolResult
import adr.spine.pure.Action
import adr.spine.pure.Actor
import adr.spine.pure.CommandId
import adr.spine.pure.Ctx
import adr.spine.pure.DECODE_FAILED
import adr.spine.pure.Registry
import adr.spine.pure.Signature
import adr.spine.pure.StagedInput

/** One finished step, from either path: the agent loop or the human surface. */
data class FinishedStep(
    val by: Actor,
    /** The ordered off-bus inputs this step consumed (5.4). Empty is the common case. */
    val staged: List<StagedInput>,
    val actions: List<Action>,
)

/**
 * The two closed maps of the boundary, over ONE registry — the symmetric pair named in
 * this file's header. Built by the Boundary from the registry it already holds.
 */
class ActionResolution<S>(private val registry: Registry<S>) {

    /**
     * The ONE closed name→ToolResult map, executed BEFORE the fold.
     *
     * A missing verb and an undecodable input both become a folded, committed
     * ToolResult.Unhandled — never a silent drop (6.5). The shipped Kotlin port's
     * `runCatching { … }.getOrNull()` is gone with it.
     */
    fun resolve(action: Action, ctx: Ctx<S>): ToolResult {
        val verb = registry[action.tool]
            ?: return ToolResult.Unhandled(action.tool, "no registered verb")
        return verb.resolve(action.input, ctx)
            ?: ToolResult.Unhandled(action.tool, DECODE_FAILED)
    }

    /**
     * The name→Command map (6.8), supplied by the same registry. EVERY verb signs —
     * presentation and domain alike (A1) — and so do the spine's own two results,
     * because a refusal is a decision and 5.4's discriminator answers yes.
     *
     * The two `is` checks are not an open match dressed up as a closed one: the spine
     * structurally CANNOT enumerate block cases (that is L1), and the registry is what
     * closes the set instead. Gate check C13 proves every case in the system has a verb
     * and signs, so the fallback is total by test.
     */
    fun sign(result: ToolResult, sig: Signature, id: CommandId): Command {
        if (result is ToolResult.Unhandled) return Command.Unhandled(result.tool, sig, id, result.note)
        if (result is ToolResult.Refused) return Command.Refused(result.tool, sig, id, result.reason)
        return registry.getValue(result.tool).signOf(result, sig, id)
    }
}

File: examples/typescript/src/spine/boundary/action.ts
// ── spine/boundary/action — the ONE name→ToolResult map (G1) ───────────────
// The human path into the fold, named at last. `fold` consumes ToolResults; a
// surface (and the agent loop) emits Actions; THIS is the conversion, and it is
// closed, boundary-owned, and executed BEFORE the fold.
//
// It is symmetric to the name→Command map (`registry[tool].sign`), which the
// SAME registry supplies. One registration, two maps — exactly what 6.8 says,
// now with both maps named.
//
// Two consequences worth stating:
//
//  * `resolveAction` is the SINGLE PRODUCTION SITE of every ToolResult in the
//    system (gate check C7), so a recorded result can never disagree with what
//    the boundary folded, and a decode failure becomes a committed `Unhandled`
//    instead of a silently dropped action.
//  * because the open-name guard lives HERE — at the boundary, where the open
//    name actually arrives — the fold has no `else` arm at all. It is
//    exhaustive over a fully closed ToolResult including Unhandled and Refused.
//    That is 6.10's "close what you own; guard what you do not" put in the right
//    place, and it is what makes the compile-time edit list total.

import type { Signature } from "../pure/actor";
import type { CommandBase, SpineCommand } from "../pure/command";
import type { CommandId, StepIndex, ToolName } from "../pure/ids";
import type { StagedInput } from "../pure/staged";
import type { Action } from "../pure/step-record";
import type { ToolResultBase } from "../pure/tool-result";
import { isSpineResult, unhandled } from "../pure/tool-result";
import type { Ctx, Verb } from "../pure/verb";

export type { Action } from "../pure/step-record";

/** One finished step, from EITHER path. Both send ACTIONS — never results — so
 *  both resolve through the one map below, and 3.2's "a person tapping a control
 *  and the agent calling a tool resolve to the identical Command" is true rather
 *  than aspirational.
 *
 *  IT CARRIES NO `Actor`, AND THAT ABSENCE IS THE INVARIANT. `by` used to be a
 *  field here, which made WHO ACTED a claim the payload made about itself: the
 *  boundary fed it verbatim to `authorityOf(step.by, session)`, so anything that
 *  could reach this seam chose its own attribution AND its own principal. That is
 *  the class `Signature` closed one layer down (spine/pure/actor), and it is
 *  closed here the same way — by making the value UNREPRESENTABLE rather than
 *  merely unused.
 *
 *  NO NAME-KEYED RULE COULD HAVE DONE IT IN THIS PORT. `Actor` is a string-literal
 *  union, so `by: "Spine"` is a bare literal in an object literal: no import, no
 *  identifier, nothing for a rule keyed on a name to see — and a hoisted
 *  `const b = "Spine"` or an `as Actor` assertion evades a rule keyed on the
 *  literal too. Deleting the field is what a lint could not reach. */
export interface FinishedStep {
  readonly staged: readonly StagedInput[];
  readonly actions: readonly Action[];
}

/** THE ACTOR RIDES THE CHANNEL, not the payload — one channel per `Actor` value,
 *  each minted by the boundary and handed to exactly one owner at wiring: the
 *  surface controller gets `human`, the agent loop and every turn get `agent`,
 *  and the serial consumer's own authored steps get `spine`.
 *
 *  This is what §5.3's "decided by where it entered, never by what it asks for"
 *  costs to make TRUE. A holder stamps what its channel stamps and nothing else,
 *  so the agent path can no longer claim `Human`, and `spine:consumer` — the one
 *  principal in the reference wiring with no credential behind it, minted to mean
 *  "no model chose this" — is reachable only from the consumer that owns it.
 *
 *  NAMED RESIDUE, because "unforgeable" would be a lie: the composition root
 *  constructs the boundary and therefore holds all three channels, exactly as it
 *  holds the authorization seam that decides what each Actor resolves to. That is
 *  the residue `spine/boundary` already has for `Signature` — the minting folder
 *  can mint. What is closed is every holder that is NOT the root: a tool, a fold
 *  arm, a turn, the surface, the agent loop. */
export interface StepChannel {
  submit(step: FinishedStep): StepIndex;
}

export type Registry<S> = ReadonlyMap<ToolName, Verb<S>>;

export function registryOf<S>(verbs: readonly Verb<S>[]): Registry<S> {
  return new Map(verbs.map((v) => [v.name, v]));
}

/** name → ToolResult. Closed, boundary-owned, pre-fold. */
export function resolveAction<S>(
  registry: Registry<S>,
  action: Action,
  ctx: Ctx<S>,
): ToolResultBase {
  const verb = registry.get(action.tool);
  if (verb === undefined) return unhandled(action.tool, "no registered verb");
  const decoded = verb.decode(action.input);
  if (!decoded.ok) return unhandled(action.tool, "input failed to decode");
  return verb.run(decoded.input, ctx);
}

/** name → Command. The other half of the same registration (6.8). Under it
 *  EVERY verb signs — presentation and domain alike — and so do the spine's own
 *  two cases, because a refusal is a decision someone may need to ask about. */
export function signResult<S>(
  registry: Registry<S>,
  result: ToolResultBase,
  sig: Signature,
  id: CommandId,
): CommandBase {
  if (isSpineResult(result)) {
    switch (result.outcome) {
      case "unhandled": {
        const cmd: SpineCommand = {
          outcome: "unhandled",
          tool: result.tool,
          sig,
          id,
          note: result.note,
        };
        return cmd;
      }
      case "refused": {
        const cmd: SpineCommand = {
          outcome: "refused",
          tool: result.tool,
          sig,
          id,
          reason: result.reason,
        };
        return cmd;
      }
      default: {
        const _never: never = result;
        return _never;
      }
    }
  }
  const verb = registry.get(result.tool);
  // Unreachable: an "ok" result can only have come out of a registered verb's
  // own `run`, six lines up. Total anyway — the spine never throws at a seam.
  if (verb === undefined) {
    const cmd: SpineCommand = {
      outcome: "unhandled",
      tool: result.tool,
      sig,
      id,
      note: "no registered verb",
    };
    return cmd;
  }
  // A COMMAND MAY ONLY CARRY THE STAMP THIS STEP MINTED — checked by IDENTITY,
  // and this is the ONLY layer that closes the constructed-forge class.
  //
  // `CommandBase.sig` (pure/command.ts) is the ONLY Signature-typed value block
  // code can emit — `ArmOut` and `FoldOut` carry none — and `verb.sign` has
  // exactly ONE call site, which is this line. So reference equality here is
  // TOTAL over "a forged stamp reaching committed transport", where no
  // type-level wall can be: `Object.assign`'s `T & U`, `structuredClone`'s
  // `T -> T`, any user-written `<T>(t: T, o: Partial<T>) => T` or
  // `<T, K extends keyof T>(t: T, k: K, v: T[K]) => T`, and
  // `new (sig.constructor as new (…) => Signature)(…)` reflection — the last of
  // which produces a REAL branded, frozen Signature, so no shape check could
  // ever catch it — all yield values the compiler accepts as `Signature` no
  // matter how the brand is spelled. Every one of them fails to be THIS object.
  //
  // Total, not a throw: the "no registered verb" fallback ten lines up is the
  // idiom to mirror, and this file's header already says the spine never throws
  // at a seam. A refusal is a decision, so it is signed — with the stamp the
  // boundary minted, never the one the verb handed back.
  const cmd = verb.sign(result, sig, id);
  if (cmd.sig !== sig) {
    const refused: SpineCommand = {
      outcome: "refused",
      tool: result.tool,
      sig,
      id,
      reason: "forged signature",
    };
    return refused;
  }
  return cmd;
}
